docs/src/references/agent/operation-workflow.md

```diff
@@ -264,26 +246,264 @@ The script exit status and output is used to determine the next step for the com
+- `cleanup` marks the terminal state of the workflow where the command has been fully processed
+  and where the original requester is expected to clean up the command retained message storing its state.
+
+#### Restart
 
 A workflow can trigger a device restart, using the builtin __restart__ action.
 
 This action is controlled by three states:
-- the *on_executing* state to which the workflow moves before the reboot is triggered
+- the *on_exec* state to which the workflow moves before the reboot is triggered
 - the *on_success* state to which the workflow resumes after a successful device reboot
 - the *on_error* state to which the workflow resumes in case the reboot fails
 
 For instance, the following triggers a reboot from the `reboot_required` state
 and moves to `restarting` waiting for the device to restart
 and finally to either `successful_restart` or `failed_restart`,
-depending on the actual status of the reboot. 
+depending on the actual status of the reboot.
+
+```toml
+["device-restart"]
+action = "restart"
+on_exec = "waiting-for-restart"
+on_success = "successful_restart"
+on_error = "failed_restart"
+```
+
+Alternative proposal:
+
+```toml
+["device-restart"]
+action = "restart"
+on_exec = "waiting-for-restart"
+
+["waiting-for-restart"]
+action = "waiting restart"
+on_success = "successful_restart"
+on_error = "failed_restart"
+```
+
+#### Waiting
+
+```toml
+["<state>"]
+action = "waiting <delegate>"
+on_success = "<successful_state>"
+on_error = "<failed_state>"
+```
+
+#### Cleanup
+
+```toml
+["successful"]
+action = "cleanup"
+
+["failed"]
+action = "cleanup"
+```
+
+#### Proceed and Builtin actions
+
+The `"proceed"` and `"builtin"` actions are useful when customizing a builtin operation
+(`software_list`, `software_update`, `restart`, `config_snapshot`, `config_update`, `log_upload`).
+Indeed, the first step is to start with a workflow specification which mimics the builtin behavior.
+
+For instance, here is the builtin workflow for the `software_update` operation:
+
+```toml
+operation = "software_update"       # an operation for which tedge-agent provides an implementation
+
+["init"]
+action = "proceed"                  # open to customization
+on_success = "scheduled"
+
+[scheduled]
+action = "builtin"                  # delegated to the tedge-agent
+on_success = "executing"
+
+[executing]
+action = "builtin"                  # delegated to the tedge-agent
+on_success = "successful"
 
+[successful]
+action = "cleanup"                  # waiting for the mapper to clean up the command
+
+[failed]
+action = "cleanup"                  # waiting for the mapper to clean up the command
 ```
-[reboot_required]
-script = "restart"
-next = ["restarting", "successful_restart", "failed_restart"]
+
+The action for the `"init"` state is a `"proceed"` action, meaning nothing specific is done by the __tedge-agent__
+and that a user can provide its own implementation.
```

Comment:
This can be changed by providing a *new* workflow definition as done here: https://github.com/thin-edge/thin-edge.io/pull/2496/files#diff-0bc2ab86d1dc663436d9764c1ee3b48333013f3d52ce410aac0af84705f0bf7cR491